scipy/stats/_hypotests.py

```diff
@@ -9,11 +9,101 @@
+    for sample in samples[1:]:
+        nans = nans | np.isnan(sample)
+    not_nans = ~nans
+    return [sample[not_nans] for sample in samples]
+
+
+def _vectorize_hypotest_factory(result_creator, default_axis=0,
+                                n_samples=1, paired=False):
+    def vectorize_hypotest_decorator(hypotest_fun_in):
+        @wraps(hypotest_fun_in)
+        def vectorize_hypotest_wrapper(*args, axis=default_axis,
+                                       nan_policy='propagate', **kwds):
+
+            # if n_samples is None, all args are samples
+            n_samp = len(args) if n_samples is None else n_samples
+
+            # split samples from other arguments
+            samples = [np.atleast_1d(sample) for sample in args[:n_samp]]
+            args = args[n_samp:]
+
+            if axis is None:
+                samples = [sample.ravel() for sample in samples]
+                axis = 0
+
+            # if axis is not needed, just handle nan_policy and return
+            ndims = np.array([sample.ndim for sample in samples])
+            if np.all(ndims <= 1):
+                # Addresses nan_policy="propagate"
+                if nan_policy == 'propagate':
+                    return hypotest_fun_in(*samples, *args, **kwds)
+
+                # Addresses nan_policy="raise"
+                contains_nans = []
+                for sample in samples:
+                    contains_nan, _ = (
+                        scipy.stats.stats._contains_nan(sample, nan_policy))
+                    contains_nans.append(contains_nan)
+
+                # Addresses nan_policy="omit" (doesn't get here otherwise)
+                if any(contains_nans):
+                    # consider passing in contains_nans
+                    samples = _remove_nans(samples, paired)
+
+                return hypotest_fun_in(*samples, *args, **kwds)
+
+            # otherwise, concatenate all samples along axis, remembering where
+            # each separate sample begins
+            lengths = np.array([sample.shape[axis] for sample in samples])
+            split_indices = np.cumsum(lengths)
+            x = scipy.stats.stats._broadcast_concatenate(samples, axis)
+
+            # Addresses nan_policy="raise"
+            contains_nan, _ = (
+                scipy.stats.stats._contains_nan(x, nan_policy))
+
+            # Addresses nan_policy="omit"
+            # Assumes nan_policy="propagate" is handled by hypotest_fun_in
+            def hypotest_fun(x):
+                samples = np.split(x, split_indices)[:n_samp]
+                if nan_policy == 'omit' and contains_nan:
+                    samples = _remove_nans(samples, paired)
+                return hypotest_fun_in(*samples, *args, **kwds)
+
+            if np.all(ndims == 1) is None:
+                return hypotest_fun(x)
+            else:
+                x = np.moveaxis(x, axis, -1)  # needed by result_creator
+                res = np.apply_along_axis(hypotest_fun, axis=-1, arr=x)
```

Comment:
In our case, as the function might be costly, maybe we could introduce `workers` and use multiprocessing instead of this?

Reply:
Maybe in a followup? This is like the "vanilla" version; `workers` would be a nice enhancement if multiple cores are available, right?

Reply:
Sure! I planned to have a look at this for bootstrap, so once this one comes in I should know how to do it quickly as well.

Reply:
Cool. Would you also make a `batch` argument (or similar) so that users can select the size of the batch in which the statistic is calculated for resamples? Currently, everything is done in one call, and that takes a ton of memory.
We should coordinate the order of operations on bootstrap improvements. @jb-leger was going to pythranize some of the resampling functions. Wouldn't want to duplicate work or cause merge conflicts if we can avoid it.

Reply:
Right we could have a batch size if the underlying function takes advantage of being feed with multiple samples 👍 
Let's see this once it's merged then. Looks like these SciPy meetups are really needed to have a discussion space 😉